src/virt.rs

```diff
@@ -314,58 +314,150 @@ impl VirtContext {
         }
     }
 
-    pub fn handle_device_access_fault(&mut self, instr: &Instr, device: &VirtDevice) {
-        match instr {
-            Instr::Ld { rd, rs1, imm } => {
-                let address = utils::calculate_addr(self.get(rs1), *imm);
-                let offset = address - device.start_addr;
-                match device.device_interface.read_device(offset) {
-                    Ok(value) => {
-                        self.set(rd, value);
-                        self.pc += 4;
-                    }
-                    Err(err) => {
-                        panic!("Error reading {}: {}", device.name, err);
-                    }
-                }
-            }
-            Instr::CLd { rd, rs1, imm } => {
-                let address = self.get(rs1) + imm * 8;
-                let offset = address - device.start_addr;
-                match device.device_interface.read_device(offset) {
-                    Ok(value) => {
-                        self.set(rd, value);
-                        self.pc += 2;
-                    }
-                    Err(err) => {
-                        panic!("Error reading {}: {}", device.name, err);
-                    }
-                }
+    // TODO: think about passing instructions instead of registers
+    // TODO: maybe join load-stores?
```

Comment:
Well, for now 2 separate loads and stores don't look too bad (maybe 30% common code overall and 50-60% between functions of the same class).
But potentially there are also float and unsigned loads and stores for the standard rv64 spec, and we might want to implement them as well.
And if the current pattern continues, it will be 4 loads and 4 stores doing mostly the same, which i believe will reduce code's readability and writability. On the other side, keepeng the handlers separate may ease the process of verification and attesting, as it would be clear what is being done for each type of loads&store and reviewing that against specification is a trivial task.

In other words, i'm unsure if the potential advantages of the join are enough to cover it's shortcomings, so counting on you to determine which approach to take here.

**P.S**: Just to clarify, i meant joining loads and stores among themselves (so like 1 general store and 1 general load ideally)

Reply:
It see, that makes sense.
I would say let's keep it like this for now and revisit the decision when we need to add floating points, the current implementation seems clear enough to me.